core/src/apps/cardano/layout.py

```diff
@@ -276,32 +277,70 @@ async def confirm_inline_datum(first_chunk: bytes, inline_datum_size: int) -> No
 async def confirm_reference_script(
     first_chunk: bytes, reference_script_size: int
 ) -> None:
-    await _confirm_data_chunk(
+    await _confirm_tx_data_chunk(
         "confirm_reference_script",
         "Reference script",
         first_chunk,
         reference_script_size,
     )
 
 
-async def _confirm_data_chunk(
-    br_type: str, title: str, first_chunk: bytes, data_size: int
+async def confirm_message_payload(
+    payload_first_chunk: bytes,
+    payload_hash: bytes,
+    payload_size: int,
+    is_signing_hash: bool,
+    prefer_hex_display: bool = False,
 ) -> None:
+    props: list[PropertyType]
+
+    if not payload_first_chunk:
+        assert payload_size == 0
+        props = _get_data_chunk_props(
+            "Empty message", payload_first_chunk, payload_size
+        )
+    elif not prefer_hex_display and is_unambiguous_ascii(payload_first_chunk):
```

Comment:
I'm wary of doing any guesswork here. Why don't we let the client choose the "display mode" (ascii/hex) explicitly and just validate it's OK?